aptos-move/framework/aptos-stdlib/sources/pool_u64.spec.move

```diff
@@ -0,0 +1,166 @@
+            vector::eq_push_back(pool.shareholders, old(pool.shareholders), shareholder);
+    }
+
+    spec fun spec_amount_to_shares_with_total_coins(pool: Pool, coins_amount: u64, total_coins: u64): u64 {
+        if (pool.total_coins == 0 || pool.total_shares == 0) {
+            coins_amount * pool.scaling_factor
+        }
+        else {
+            (coins_amount * pool.total_shares) / total_coins
+        }
+    }
+
+    spec amount_to_shares_with_total_coins(pool: &Pool, coins_amount: u64, total_coins: u64): u64 {
+        aborts_if pool.total_coins > 0 && pool.total_shares > 0
+            && (coins_amount * pool.total_shares) / total_coins > MAX_U64;
+        aborts_if (pool.total_coins == 0 || pool.total_shares == 0)
+            && coins_amount * pool.scaling_factor > MAX_U64;
+        aborts_if pool.total_coins > 0 && pool.total_shares > 0 && total_coins == 0;
+        ensures result == spec_amount_to_shares_with_total_coins(pool, coins_amount, total_coins);
+    }
+
+    spec shares_to_amount_with_total_coins(pool: &Pool, shares: u64, total_coins: u64): u64 {
+        aborts_if pool.total_coins > 0 && pool.total_shares > 0
+            && (shares * total_coins) / pool.total_shares > MAX_U64;
+        ensures result == spec_shares_to_amount_with_total_coins(pool, shares, total_coins);
+    }
+
+    spec fun spec_shares_to_amount_with_total_coins(pool: Pool, shares: u64, total_coins: u64): u64 {
+        if (pool.total_coins == 0 || pool.total_shares == 0) {
+            0
+        }
+        else {
+            (shares * total_coins) / pool.total_shares
+        }
+    }
+
+    spec multiply_then_divide(_pool: &Pool, x: u64, y: u64, z: u64): u64 {
+        aborts_if z == 0;
+        aborts_if (x * y) / z > MAX_U64;
+        ensures result == (x * y) / z;
+    }
+
+    spec redeem_shares(pool: &mut Pool, shareholder: address, shares_to_redeem: u64): u64 {
+        let redeemed_coins = spec_shares_to_amount_with_total_coins(pool, shares_to_redeem, pool.total_coins);
+        aborts_if !spec_contains(pool, shareholder);
+        aborts_if spec_shares(pool, shareholder) < shares_to_redeem;
+        aborts_if pool.total_coins < redeemed_coins;
+        aborts_if pool.total_shares < shares_to_redeem;
+        ensures pool.total_coins == old(pool.total_coins) - redeemed_coins;
+        ensures pool.total_shares == old(pool.total_shares) - shares_to_redeem;
+        include shares_to_redeem > 0 ==> DeductSharesEnsures { num_shares: shares_to_redeem };
+        ensures result == redeemed_coins;
+    }
+
+    spec transfer_shares(
+        pool: &mut Pool,
+        shareholder_1: address,
+        shareholder_2: address,
+        shares_to_transfer: u64
+    ) {
+        pragma aborts_if_is_partial;
+        aborts_if !spec_contains(pool, shareholder_1);
+        aborts_if spec_shares(pool, shareholder_1) < shares_to_transfer;
+        // TODO: difficult to specify due to the intermediate state problem.
```

Comment:
What intermediate state problem? Can you elaborate? 🙏

Reply:
The contract seems to have no issue. The intermediate state problem is a current limitation in the spec method. In this specific example, `deduct_shares` and `add_shares` are individually verified. However, `transfer_shares` calls `deduct_shares` and `add_shares` in a row, and so there is an implicit intermediate state between the tow function calls. Prover currently has no way to refer to the intermediate state in the spec although the precondition of `add_shares` should be based on the intermediate state.